Wrappers/Python/ccpi/viewer/utils/io.py

```diff
@@ -42,3 +66,765 @@ def SaveRenderToPNG(render_window, filename):
+
+            if file_extension in ['.mha', '.mhd']:
+                reader = self._GetMetaImageReader()
+
+            elif file_extension in ['.npy']:
+                reader = self._GetNumpyImageReader()
+
+            elif file_extension in ['.raw']:
+                reader = self._GetRawImageReader()
+
+            elif file_extension in ['.nxs', '.h5', '.hdf5']:
+                reader = self._GetHDF5ImageReader()
+                self._OriginalImageAttrs['dataset_name'] = self._HDF5DatasetName
+
+            elif file_extension in ['.tif', '.tiff']:
+                image_files = glob.glob(os.path.join(os.path.dirname(self._FileName), '*.{}'.format(file_extension)))
+                if len(image_files) == 0:
+                    raise Exception('No tiff files were found in: {}'.format(self._FileName))
+                self._data = self._GetTiffImageReader(image_files)
+
+            else:
+                raise Exception('File format is not supported. Accepted formats include: .mhd, .mha, .npy, .tif, .raw')
+        else:  # If we are given a folder, not a file, look for tiff files and try to read them
+            image_files = glob.glob(os.path.join(self._FileName, '*.tif')) + glob.glob(
+                os.path.join(self._FileName, '*.tiff'))
+            if len(image_files) == 0:
+                raise Exception('No tiff files were found in: {}'.format(self._FileName))
+            self._data = self._GetTiffImageReader(image_files)
+
+        if file_extension not in ['.tif', '.tiff']:
+            # currently the tiff reader doesn't take these inputs:
+            reader.SetFileName(self._FileName)
+            # if we have acquisition data then we would not resample on
+            # the z axis:
+            reader.SetIsAcquisitionData(not self._ResampleZ)
+        if not self._Crop:
+            if self._Resample:
+                target_size = self._TargetSize
+            else:
+                # forced use of resample reader in the case that we
+                # don't want to crop or resample,
+                # but the large target size means we don't resample
+                target_size = 1e12
+            reader.SetTargetSize(int(target_size))
+
+        # Add observers:
+        reader.AddObserver(vtk.vtkCommand.ProgressEvent,
+                           partial(self._ReportProgress, progress_callback=progress_callback))
+
+        # Prints the error if an error occurs in the reader.
+        # Otherwise this wouldn't print at all.
+        error_obs = ErrorObserver(callback_fn=print)
+        reader.AddObserver(vtk.vtkCommand.ErrorEvent, error_obs)
+        # Could add end observer so that we don't continue to do anything
+        # else if an error does occur in reader?
+
+        return reader
+
+    def _GetMetaImageReader(self, progress_callback=None):
+        if self._Crop:
+            reader = cilMetaImageCroppedReader()
+            reader.SetTargetZExtent(tuple(self._TargetZExtent))
```

Comment:
all readers, with the https://github.com/vais-ral/CILViewer/blob/4cc7472bb0e11a79f6ae163894f66068568b6a59/Wrappers/Python/ccpi/viewer/utils/conversion.py#L1900

support `SetTargetZExtent` which could be moved in the `Read` method, at about line 315

Reply:
I didn't do it for this reason: https://github.com/vais-ral/CILViewer/pull/216/files/0944550751358d7549db2a3b8ebadc4877d6c303#r944597190